### ISSUE TITLE
Demo/feature suggestion: Registering 2+ encoders and handling them gracefully in single callbacks

### ISSUE THREAD

**Issue by Ryder17z** [Stutchbury/EncoderButton#1]
Hello. This is a very nice library because it actually works reliably with Teensy and KY-040 encoders.

However, having two or more encoders connected does lead to a bit of a mess.

I'd like to purpose a solution for this, like the psudo-code below.

```
/**
 * An example of using two encoders with the EncoderButton library.
 */
#include <EncoderButton.h>

EncoderButton eb1(12, 11, 10); // assigned ID 0 by default
EncoderButton eb2(9, 8, 7); // assigned ID 1 automatically

void onEbClicked(EncoderButton& eb) {
  if(eb.id == 0)
  {
    Serial.print("eb1 clickCount: ");
    Serial.println(eb.clickCount);
    eb.resetPosition(); // example use-case, reset it it's own position
  }
  else if(eb.id == 1)
  {
    Serial.print("eb2 clickCount: ");
    Serial.println(eb.clickCount);

    // example use-case, reset both encoders
    eb1.resetPosition();
    eb2.resetPosition(); 
  }
}

/**
 * A function to handle the 'encoder' event
 */
void onEbEncoder(EncoderButton& eb) {
  if(eb.id == 0)
  {
    Serial.print("eb1 incremented by: ");
    Serial.println(eb.increment());
    Serial.print("eb1 position is: ");
    Serial.println(eb.position());
  }
  else if(eb.id == 1)
  {
    Serial.print("eb2 incremented by: ");
    Serial.println(eb.increment());
    Serial.print("eb2 position is: ");
    Serial.println(eb.position());
  }
}

void setup() {
  // put your setup code here, to run once:
  Serial.begin(9600);
  delay(500);
  Serial.println("EncoderButton 2-encoder Example");

  //Link the event(s) to your function
  eb1.setClickHandler(onEbClicked);
  eb1.setEncoderHandler(onEbEncoder);

  eb2.setClickHandler(onEbClicked);
  eb2.setEncoderHandler(onEbEncoder);
}

void loop() {
  eb.update(); // global updater
}
```

**Comment by Stutchbury** [comment]
Hi - Thank you for you feedback - it is much appreciated. The use of multiple encoders should not create a mess as you would normally create a separate callback function for each encoder (indeed, for each event of each encoder) - it is unusual to have a common callback function (idle timeout being the common exception).

Having said that,  I currently have an update waiting to commit that will provide both a userID and a userState (both int) for the encoder/button - but it is set by the integrator (user), not automatically. In [this project ](https://forum.linuxcnc.org/18-computer/44682-arduino-teensy-python-interface-manualmatic-mpg-pendant) there are three encoders (one an MPG) and eleven buttons - five of which are multi-function and share a single onClickHandler - hence the need for an id.

It is highly unlikely I'll go down the route of a common/singleton update() - the management of each individual encoder/button instance should be left to the integrator.

I'll create a multi-button and a shared-callback example before I push the userID/State update.

**Comment by ExileFox** [comment]
`It is highly unlikely I'll go down the route of a common/singleton update() - the management of each individual encoder/button instance should be left to the integrator.`

I guess it boils down to the libraries/frameworks involved. I'm very used to seeing a couple or a single main callback(s) so maybe I've gotten a bit biased. Do what you feel is best. 
 


`I'll create a multi-button and a shared-callback example before I push the userID/State update.`

That sounds great.

**Comment by Stutchbury** [comment]
> I'm very used to seeing a couple or a single main callback(s)

I had not really considered shared callbacks for anything other than across encoders/buttons - sharing across events would require identifying the event too. Could be interesting but I'm wary of straying to far :-)

> That sounds great.

I've pushed the userId/userState changes and the examples - I hope you find this useful. I've also bumped the release so it should appear in the Arduino Library Manager soon.
